### PR TITLE
Add table/layout tests

### LINKED ISSUE #283: Cloned ScrollableControl.DockPaddingEdges.Top actually equals the Bottom!
```cs

[Fact]
public void Clone_InvokeWithOwner_ReturnsExpected()
{
    var owner = new ScrollableControl
    {
        Padding = new Padding(1, 2, 3, 4)
    };
    ICloneable original = owner.DockPadding;
    ScrollableControl.DockPaddingEdges padding = (ScrollableControl.DockPaddingEdges)original.Clone();
    Assert.Equal(1, padding.Left);
    Assert.Equal(2, padding.Top); // FAILS: this is set to 4!!
    Assert.Equal(3, padding.Right);
    Assert.Equal(4, padding.Bottom);
}
```

### LINKED ISSUE #284: ScrollableControl.DockPaddingEdges.Equals for a cloned padding throws NRE
```cs
[Fact]
public void Equals_ClonedThis_ReturnsExpected()
{
    var owner = new ScrollableControl
    {
        Padding = new Padding(1, 2, 3, 4)
    };
    ICloneable original = owner.DockPadding;
    ScrollableControl.DockPaddingEdges padding = (ScrollableControl.DockPaddingEdges)original.Clone();
    Assert.Throws<NullReferenceException>(() => padding.Equals(original));
}

[Fact]
public void Equals_ClonedOther_ReturnsExpected()
{
    var owner = new ScrollableControl
    {
        Padding = new Padding(1, 2, 3, 4)
    };
    ICloneable original = owner.DockPadding;
    ScrollableControl.DockPaddingEdges padding = (ScrollableControl.DockPaddingEdges)original.Clone();
    Assert.Throws<NullReferenceException>(() => original.Equals(padding));
}
```

### LINKED ISSUE #290: ScrollableControl.DockPaddingEdges.GetHashCode returns different values for equal
Overrides `base.GetHashCode` which returns different results for the same padding but with a different owner. Also means that an equal clone does not have the same hash code as the original

### LINKED ISSUE #291: ScrollableControl.DockPaddingEdges.ToString returns ""
Should return something descriptive at least, e.g. matching `Padding`

### PR DESCRIPTION
**Pull request by hughbe**
And fix bad debug asserts and fix #283 and fix #284 
Fixes #290 and Fixes #291

### CODE REVIEW

**Comment by zsd4yr** [comment]
Also noticing that you could be using Mocks via Moq in a large variety of cases instead of extending your own class just for testing. This is our preferred method

File: src/System.Windows.Forms/src/System/Windows/Forms/ScrollableControl.cs [hunk 1]
```diff
@@ -1679,21 +1679,21 @@ public class DockPaddingEdges : ICloneable {
...
+                    Left == dpeOther.Left &&
+                    Top == dpeOther.Top &&
+                    Right == dpeOther.Right &&
+                    Bottom == dpeOther.Bottom;
             }
 
-            /// <include file='doc\ScrollableControl.uex' path='docs/doc[@for="ScrollableControl.DockPaddingEdges.GetHashCode"]/*' />
-            /// <internalonly/>
-            public override int GetHashCode() {
-                return base.GetHashCode();
```
[the remark this answers is not in the document]
Comment (hughbe) [reply src/System.Windows.Forms/src/System/Windows/Forms/ScrollableControl.cs]: Is this due to compat concerns? The main reason I changed the implementation of `GetHashCode` is because it is non-deterministic, as it relies on the value of the pointer `Owner`. This meant that a cloned `DockPaddingEdges` would not have the same hash code even though its equal to the original. That seemed strange to me and doesn't following hashing guidelines

**Comment by hughbe** [comment]
> add ClassName_ to the front of all these tests. I realize this is tedious but it really makes looking at the results much easier

I'm not sure I agree with this. The test class itself describes what we're testing

Take the following example. It's fairly obvious that there's a failure in `MenuItem` tests. In fact, surely adding another suffix `MenuItemTests.MenuItem_Ctor_String_MenuItemArray` actually makes it harder to see what failed and where?
```
[xUnit.net 00:00:01.29]     System.Windows.Forms.MenuItemTests.Ctor_String_MenuItemArray(text: "value", items: [System.Windows.Forms.MenuItem, Items.Count: 0, Text: ]) [FAIL]
Failed   System.Windows.Forms.MenuItemTests.Ctor_String_MenuItemArray(text: "value", items: [System.Windows.Forms.MenuItem, Items.Count: 0, Text: ])
Error Message:
 Assert.False() Failure
Expected: False
Actual:   True
Stack Trace:
   at System.Windows.Forms.MenuItemTests.Ctor_String_MenuItemArray(String text, MenuItem[] items) in /Users/hugh/Documents/GitHub/winforms/src/System.Windows.Forms.Design/tests/System/Windows/Forms/MenuItemTests.cs:line 199
```

In terms of the other changes, I'm not free for the next couple of days on holiday but I agree some stuff can be moved to `TestHelpers` or `Moq` and I'll remove or add parameters for the ternaries - I don't want to seem to be testing the implementation by simply copying conditions into the tests!

File: src/System.Windows.Forms/src/System/Windows/Forms/ScrollableControl.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (zsd4yr) [reply src/System.Windows.Forms/src/System/Windows/Forms/ScrollableControl.cs]: I believe the base _is_ deterministic _within each session_, which is sufficient for the purposes of single-session hashing. If, however, your concern is accessing this data across sessions, I can understand the desire to change the function.

Alternatively, if you disagree with my premise, we should probably discuss globally changing these hash code overloads to be deterministic across sessions.

**Comment by zsd4yr** [comment]
> In terms of the other changes, I'm not free for the next couple of days on holiday but I agree some stuff can be moved to `TestHelpers` or `Moq` and I'll remove or add parameters for the ternaries - I don't want to seem to be testing the implementation by simply copying conditions into the tests!

Sounds great 😄 enjoy your holiday, and I'll eagerly await the remaining requested changes.

File: src/System.Windows.Forms/src/System/Windows/Forms/ScrollableControl.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (hughbe) [reply src/System.Windows.Forms/src/System/Windows/Forms/ScrollableControl.cs]: Shall I open an issue?

File: src/System.Windows.Forms/src/System/Windows/Forms/ScrollableControl.cs [hunk 2]
```diff
@@ -1661,7 +1661,7 @@ public class DockPaddingEdges : ICloneable {
             public int Top {
                  get { 
                     if (owner == null) {
-                        return bottom;
+                        return top;
```
Comment (sharwell) [inline src/System.Windows.Forms/src/System/Windows/Forms/ScrollableControl.cs]: ❓ Can we create bugs for all behavior issues identified during testing? I don't want the production code improvements to get lost among the noise.

File: src/System.Windows.Forms/src/System/Windows/Forms/ScrollableControl.cs [hunk 3]
```diff
@@ -1726,10 +1726,9 @@ public class DockPaddingEdges : ICloneable {
                 this.owner.Padding.Scale(dx, dy);
             }
 
-            /// <include file='doc\ScrollableControl.uex' path='docs/doc[@for="ScrollableControl.DockPaddingEdges.ToString"]/*' />
-            /// <internalonly/>
-            public override string ToString() {
-                return "";      // used to say "(DockPadding)" but that's useless
+            public override string ToString()
+            {
+                return $"{{Left={Left},Top={Top},Right={Right},Bottom={Bottom}}}";
```
Comment (sharwell) [inline src/System.Windows.Forms/src/System/Windows/Forms/ScrollableControl.cs]: ❓ Why not `return owner.Padding.ToString();`?

File: src/System.Windows.Forms/src/System/Windows/Forms/ScrollableControl.cs (hunk 1 above)
Comment (sharwell) [inline src/System.Windows.Forms/src/System/Windows/Forms/ScrollableControl.cs]: ❓ Why not `return owner.Padding.GetHashCode();`?

File: src/System.Windows.Forms/src/System/Windows/Forms/TableLayoutSettings.cs [hunk 4]
```diff
@@ -208,7 +208,6 @@ public sealed class TableLayoutSettings : LayoutSettings, ISerializable {
         internal bool IsStub {
             get {
                 if ( _stub != null) {
-                    Debug.Assert( _stub.IsValid, "seems like we're still partying on an object that's given over its rows and columns, that's a nono.");
```
Comment (sharwell) [inline src/System.Windows.Forms/src/System/Windows/Forms/TableLayoutSettings.cs]: 📝 I would prefer to keep these assertions unless they were failing somehow. If they were failing, can you add some comments in this diff explaining why they were failing?

File: src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/CreateParamsTests.cs [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+        {
+            var createParams = new CreateParams
+            {
+                Height = value
+            };
+            Assert.Equal(value, createParams.Height);
+        }
+
+        [Theory]
+        [MemberData(nameof(CommonTestHelper.GetIntPtrTheoryData), MemberType = typeof(CommonTestHelper))]
```
Comment (sharwell) [inline src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/CreateParamsTests.cs]: 💭 If `MemberDataAttribute` is not sealed, we might be able to make a derived attribute `CommonMemberData` and avoid the need to set `MemberType` at each usage site.

File: src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/DockPaddingEdgesConverterTests.cs [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+        public void GetProperties_Invoke_ReturnsExpected()
+        {
+            var converter = new ScrollableControl.DockPaddingEdgesConverter();
+            PropertyDescriptorCollection properties = converter.GetProperties(null);
+            Assert.Equal(5, properties.Count);
+            Assert.Equal("All", properties[0].Name);
+            Assert.Equal("Left", properties[1].Name);
+            Assert.Equal("Top", properties[2].Name);
+            Assert.Equal("Right", properties[3].Name);
+            Assert.Equal("Bottom", properties[4].Name);
```
Comment (sharwell) [inline src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/DockPaddingEdgesConverterTests.cs]: 💡 I believe these assertions could use `nameof` for the expected values.

File: src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/Layout/TableLayoutSettingsTests.cs [hunk 7]
```diff
@@ -0,0 +1,762 @@
...
+        [MemberData(nameof(EmptySettings_TestData))]
+        public void RowStyles_Get_ReturnsExpected(TableLayoutSettings settings)
+        {
+            Assert.Same(settings.RowStyles, settings.RowStyles);
+        }
+
+        [Theory]
+        [InlineData(0)]
+        [InlineData(1)]
+        public void ColumnCount_Set_GetReturnsExpected(int value)
```
[the remark this answers is not in the document]
Comment (hughbe) [reply src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/Layout/TableLayoutSettingsTests.cs]: I tried using that data, but I actually can't use that data here, or for `RowCount` (and in `TableLayoutPanel`), as this allocates an array, and values like `int.MaxValue` can throw an OOM or slow down the tests stupidly much

File: src/System.Windows.Forms/src/System/Windows/Forms/ScrollableControl.cs (hunk 3 above)
Comment (hughbe) [reply src/System.Windows.Forms/src/System/Windows/Forms/ScrollableControl.cs]: Because `owner` can be null, for example if the edges are cloned

File: src/System.Windows.Forms/src/System/Windows/Forms/ScrollableControl.cs (hunk 1 above)
Comment (hughbe) [reply src/System.Windows.Forms/src/System/Windows/Forms/ScrollableControl.cs]: Because owner can be null, for example if the edges are cloned

File: src/System.Windows.Forms/src/System/Windows/Forms/TableLayoutSettings.cs (hunk 4 above)
Comment (hughbe) [reply src/System.Windows.Forms/src/System/Windows/Forms/TableLayoutSettings.cs]: They were failing. This can occur, for example when setting `TableLayoutPanel.LayoutSettings` which applies the new settings to the existing. Then if we use any properties on the new settings, we would hit this assertion

File: src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/CreateParamsTests.cs (hunk 5 above)
Comment (hughbe) [reply src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/CreateParamsTests.cs]: It is sealed sadly

File: src/System.Windows.Forms/src/System/Windows/Forms/ScrollableControl.cs (hunk 2 above)
Comment (hughbe) [reply src/System.Windows.Forms/src/System/Windows/Forms/ScrollableControl.cs]: Cloned ScrollableControl.DockPaddingEdges.Top actually equals the Bottom: https://github.com/dotnet/winforms/issues/283
ScrollableControl.DockPaddingEdges.Equals for a cloned padding throws NRE: ScrollableControl.DockPaddingEdges.GetHashCode returns different values for equal : https://github.com/dotnet/winforms/issues/284
https://github.com/dotnet/winforms/issues/290
ScrollableControl.DockPaddingEdges.ToString returns "" : https://github.com/dotnet/winforms/issues/291

File: src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/DockPaddingEdgesConverterTests.cs (hunk 6 above)
Comment (hughbe) [reply src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/DockPaddingEdgesConverterTests.cs]: nice

File: src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/Layout/LayoutEngineTests.cs [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+namespace System.Windows.Forms.Layout.Tests
+{
+    public class LayoutEngineTests
+    {
+        [Fact]
+        public void InitLayout_ValidChild_Nop()
+        {
+            var engine = new SubLayoutEngine();
+            engine.InitLayout(new ScrollableControl(), BoundsSpecified.All);
+        }
```
[the remark this answers is not in the document]
Comment (hughbe) [reply src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/Layout/LayoutEngineTests.cs]: No it is by definition a `nop` in base classes

File: src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/Layout/LayoutEngineTests.cs [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+
+        [Fact]
+        public void InitLayout_InvalidChild_Nop()
+        {
+            var engine = new SubLayoutEngine();
+            engine.InitLayout("child", BoundsSpecified.All);
```
[the remark this answers is not in the document]
Comment (hughbe) [reply src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/Layout/LayoutEngineTests.cs]: No it is by definition a nop in base classes

File: src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/Layout/LayoutEngineTests.cs [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+        {
+            var engine = new SubLayoutEngine();
+            Assert.False(engine.Layout(new ScrollableControl(), new LayoutEventArgs(new Component(), "affectedProperty")));
+        }
+
+        [Fact]
+        public void Layout_InvalidContainer_Nop()
+        {
+            var engine = new SubLayoutEngine();
+            engine.Layout("container", new LayoutEventArgs(new Component(), "affectedProperty"));
```
[the remark this answers is not in the document]
Comment (hughbe) [reply src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/Layout/LayoutEngineTests.cs]: No it is by definition a nop in base classes

File: src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/Layout/TableLayoutSettingsTests.cs [hunk 11]
```diff
@@ -0,0 +1,762 @@
...
+            Assert.Equal(TableLayoutPanelGrowStyle.AddRows, settings.GrowStyle);
+        }
+
+        [Fact]
+        public void Properties_GetWithoutOwner_ReturnsExpected()
+        {
+            var converter = new TableLayoutSettingsTypeConverter();
+            TableLayoutSettings settings = Assert.IsType<TableLayoutSettings>(converter.ConvertFrom(@"<?xml version=""1.0"" encoding=""utf-16""?><Root />"));
+            Assert.NotNull(settings.LayoutEngine);
+            Assert.Same(settings.LayoutEngine, settings.LayoutEngine);
```
[the remark this answers is not in the document]
Comment (hughbe) [reply src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/Layout/TableLayoutSettingsTests.cs]: The instance is cached and make sure it returns same instance

File: src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/Layout/TableLayoutSettingsTests.cs [hunk 12]
```diff
@@ -0,0 +1,762 @@
...
+            {
+                yield return new object[] { settings };
+            }
+        }
+
+        [Theory]
+        [MemberData(nameof(EmptySettings_TestData))]
+        public void ColumnStyles_Get_ReturnsExpected(TableLayoutSettings settings)
+        {
+            Assert.Same(settings.ColumnStyles, settings.ColumnStyles);
```
[the remark this answers is not in the document]
Comment (hughbe) [reply src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/Layout/TableLayoutSettingsTests.cs]: The instance is cached and make sure it returns same instance

File: src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/Layout/TableLayoutSettingsTests.cs [hunk 13]
```diff
@@ -0,0 +1,762 @@
...
+            var toolStrip = new ToolStrip { LayoutStyle = ToolStripLayoutStyle.Table };
+            TableLayoutSettings settings = Assert.IsType<TableLayoutSettings>(toolStrip.LayoutSettings);
+            Assert.Throws<NullReferenceException>(() => settings.SetColumnSpan(null, 1));
+        }
+
+        [Fact]
+        public void SetColumnSpan_NullControlStub_GetReturnsExpected()
+        {
+            var converter = new TableLayoutSettingsTypeConverter();
+            TableLayoutSettings settings = Assert.IsType<TableLayoutSettings>(converter.ConvertFrom(@"<?xml version=""1.0"" encoding=""utf-16""?><Root />"));
```
[the remark this answers is not in the document]
Comment (hughbe) [reply src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/Layout/TableLayoutSettingsTests.cs]: Slightly confused, how can this be turned into a mock? We're actually testing the result of calling `new TableLayoutSettingsTypeConverter.ConvertTo` because it has a certain different behaviour than if the settings were attached onto a control

File: src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/Layout/TableLayoutSettingsTypeConverterTests.cs [hunk 14]
```diff
@@ -0,0 +1,348 @@
...
+        public void ConvertFrom_Empty_ReturnsExpected(string value)
+        {
+            var converter = new TableLayoutSettingsTypeConverter();
+            TableLayoutSettings settings = Assert.IsType<TableLayoutSettings>(converter.ConvertFrom(value));
+            Assert.Empty(settings.ColumnStyles);
+            Assert.Empty(settings.RowStyles);
+        }
+
+        [Theory]
+        [InlineData("")]
```
[the remark this answers is not in the document]
Comment (hughbe) [reply src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/Layout/TableLayoutSettingsTypeConverterTests.cs]: `string.Empty` isn't supported in `InlineData` and really not worth the hassle of moving this to `MemberData` imo

File: src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/MessageTests.cs [hunk 15]
```diff
@@ -0,0 +1,766 @@
...
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Generic;
+using Xunit;
+
+namespace System.Windows.Forms.Tests
+{
+    public class MessageTests
+    {
+        public static IEnumerable<object[]> IntPtr_TestData()
```
[the remark this answers is not in the document]
Comment (hughbe) [reply src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/MessageTests.cs]: Comment on outdated diff. This now reads as`[MemberData(nameof(CommonTestHelper.GetIntPtrTheoryData), MemberType = typeof(CommonTestHelper))]`

File: src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/Layout/TableLayoutSettingsTypeConverterTests.cs [hunk 16]
```diff
@@ -0,0 +1,348 @@
...
+            {
+            }
+
+            public override ICustomTypeDescriptor GetTypeDescriptor(Type objectType, object instance)
+            {
+                return new CustomTypeDescriptor();
+            }
+        }
+
+        private class CustomTypeDescriptor : ICustomTypeDescriptor
```
[the remark this answers is not in the document]
Comment (hughbe) [reply src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/Layout/TableLayoutSettingsTypeConverterTests.cs]: fixed :)

File: src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/PaddingConverterTests.cs [hunk 17]
```diff
@@ -0,0 +1,316 @@
...
+        [Theory]
+        [InlineData(typeof(Padding))]
+        [InlineData(typeof(int))]
+        public void ConvertTo_InvalidDestinationType_ThrowsNotSupportedException(Type destinationType)
+        {
+            var converter = new PaddingConverter();
+            Assert.Throws<NotSupportedException>(() => converter.ConvertTo(new Padding(), destinationType));
+        }
+
+        public static IEnumerable<object[]> CreateInstance_TestData()
```
[the remark this answers is not in the document]
Comment (hughbe) [reply src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/PaddingConverterTests.cs]: The test params are slightly different:
```
        public static IEnumerable<object[]> CreateInstance_TestData()
        {
            yield return new object[] { new Padding(1), 1, new Padding(1, 2, 3, 4) };
            yield return new object[] { new Padding(1), 2, new Padding(2) };
            yield return new object[] { new Padding(1, 2, 3, 4), -1, new Padding(1, 2, 3, 4) };
            yield return new object[] { new Padding(1, 2, 3, 4), 1, new Padding(1) };
        }

        [Theory]
        [MemberData(nameof(CreateInstance_TestData))]
        public void CreateInstance_ValidPropertyValuesAll_ReturnsExpected(Padding instance, int all, Padding result)
        {
```
The padding test helper on the other hand returns
```cs
        public static TheoryData<Padding> GetPaddingTheoryData()
        {
            var data = new TheoryData<Padding>();
            data.Add(new Padding());
            data.Add(new Padding(0));
            data.Add(new Padding(1));
            data.Add(new Padding(int.MaxValue));
            return data;
        }

```

File: src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/PaddingConverterTests.cs [hunk 18]
```diff
@@ -0,0 +1,316 @@
...
+            var converter = new PaddingConverter();
+            Assert.True(converter.GetPropertiesSupported());
+        }
+
+        private class ClassWithPadding
+        {
+            public Padding Padding { get; set; }
+        }
+
+        private class CustomTypeDescriptorContext : ITypeDescriptorContext
```
[the remark this answers is not in the document]
Comment (hughbe) [reply src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/PaddingConverterTests.cs]: I actually love moq, fixed

File: src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/PanelTests.cs [hunk 19]
```diff
@@ -0,0 +1,204 @@
...
+        {
+            var panel = new Panel
+            {
+                AutoSizeMode = value
+            };
+            Assert.Equal(value, panel.AutoSizeMode);
+        }
+
+        [Theory]
+        [InlineData((AutoSizeMode)(AutoSizeMode.GrowAndShrink - 1))]
```
[the remark this answers is not in the document]
Comment (hughbe) [reply src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/PanelTests.cs]: fixed

File: src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/MessageTests.cs [hunk 20]
```diff
@@ -0,0 +1,766 @@
...
+
+        [Fact]
+        public void GetHashCode_Invoke_ReturnsExpected()
+        {
+            Message message = Message.Create((IntPtr)1, 1, (IntPtr)1, (IntPtr)1);
+            Assert.NotEqual(0, message.GetHashCode());
+            Assert.Equal(message.GetHashCode(), message.GetHashCode());
+        }
+
+        public static IEnumerable<object[]> ToString_TestData()
```
[the remark this answers is not in the document]
Comment (hughbe) [reply src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/MessageTests.cs]: Updated this PR to remove `MessageTests` and will submit as a separate PR

File: src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/Layout/TableLayoutSettingsTypeConverterTests.cs [hunk 21]
```diff
@@ -311,38 +312,8 @@ public CustomTypeDescriptionProvider()
 
             public override ICustomTypeDescriptor GetTypeDescriptor(Type objectType, object instance)
             {
-                return new CustomTypeDescriptor();
+                return Mock.Of<ICustomTypeDescriptor>(c => c.GetProperties() == new PropertyDescriptorCollection(new PropertyDescriptor[0]));
```
Comment (sharwell) [inline src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/Layout/TableLayoutSettingsTypeConverterTests.cs]: ❕ Do not use non-strict mocks (which unfortunately also eliminates the use of `Mock.Of<T>`)

**Comment by hughbe** [comment]
@zsd4yr sure - it makes more sense considering this PR is getting big and the comments are basically untrackable!

File: src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/CreateParamsTests.cs (hunk 5 above)
Comment (hughbe) [reply src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/CreateParamsTests.cs]: I've created a custom `CommonMemberDataAttribute` that sort-of reimplements `MemberDataAttribute`, inheriting from `MemberDataAttributeBase`

**Comment by zsd4yr** [comment]
Let's be sure to rerun the CI once those tests get in

File: src/Common/tests/CommonMemberDataAttribute.cs [hunk 22]
```diff
@@ -0,0 +1,36 @@
...
+
+namespace WinForms.Common.Tests
+{
+    /// <summary>
+    /// A custom MemberData attribute that is specialized for the CommonTestHelper type.
+    /// Useful to remove the need to suffix all attributes with "MemberType = ..."
+    /// We cannot inherit from MemberDataAttribute as it is sealed, so we have to reimplement
+    /// ConvertDataItem inheriting from MemberDataAttributeBase.
+    /// </summary>
+    public class CommonMemberDataAttribute : MemberDataAttributeBase
```
Comment (sharwell) [inline src/Common/tests/CommonMemberDataAttribute.cs]: Nice, this worked every bit as well as I was hoping.

💡 `sealed`

**Comment by hughbe** [comment]
https://github.com/dotnet/winforms/pull/293 needs to be merged first :)

**Comment by zsd4yr** [comment]
@hughbe done

**Comment by zsd4yr** [comment]
@AdamYoblick can you check out these tests?

File: src/System.Windows.Forms/tests/UnitTests/Panel.cs [hunk 23]
```diff
@@ -1,20 +0,0 @@
-﻿// Licensed to the .NET Foundation under one or more agreements.
-// The .NET Foundation licenses this file to you under the MIT license.
-// See the LICENSE file in the project root for more information.
-
-using Xunit;
-
-namespace System.Windows.Forms.Tests
```
Comment (zsd4yr) [inline src/System.Windows.Forms/tests/UnitTests/Panel.cs]: why was this deleted?

Comment (zsd4yr) [reply src/System.Windows.Forms/tests/UnitTests/Panel.cs]: works for me

**Comment by zsd4yr** [comment]
:shipit: thanks for all the work @hughbe
